### PR TITLE
#206 - Added delete address function 

### PR DESCRIPTION
**Pull request by ahmed-shehata**
Issue #206 = Added delete address function + fixed small bug in refreshing view state in checkout summary

### CODE REVIEW

File: AtlasUI/AtlasUI/CheckoutSummaryViewController.swift [hunk 1]
```diff
@@ -107,6 +107,10 @@ extension CheckoutSummaryViewController {
 
 extension CheckoutSummaryViewController {
 
+    func refreshViewData() {
+        setupViewState()
+    }
+
```
Comment (ahmed-shehata) [inline AtlasUI/AtlasUI/CheckoutSummaryViewController.swift]: I created a separate function for this just to encapsulate whatever is needed to refresh the view data

File: AtlasUI/AtlasUI/AddressListViewController.swift [hunk 2]
```diff
@@ -48,12 +50,38 @@ final class AddressListViewController: UIViewController, UITableViewDelegate, UI
         tableView.registerReusableCell(AddressRowViewCell.self)
```
Comment (bonkey) [inline AtlasUI/AtlasUI/AddressListViewController.swift]: You could update those 4 `... constraintEqualToAnchor...` to @hani-ibrahim's shortcut functions (and use them always)

File: AtlasUI/AtlasUI/AddressListViewController.swift [hunk 3]
```diff
@@ -48,12 +50,38 @@ final class AddressListViewController: UIViewController, UITableViewDelegate, UI
...
+        forRowAtIndexPath indexPath: NSIndexPath) {
+            if editingStyle == .Delete {
+                let address = self.addresses[indexPath.item]
+                checkout.client.deleteAddress(address.id) { result in
+                    switch result {
+                    case .success(_):
+                        self.addresses.removeAtIndex(indexPath.item)
+                        tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
+                    case .failure(let error):
+                        AtlasLogger.logError(error)
```
Comment (bonkey) [inline AtlasUI/AtlasUI/AddressListViewController.swift]: no `self.userMessage.show(error:)`?

File: AtlasUI/AtlasUI/AddressPickerViewController.swift [hunk 4]
```diff
@@ -72,7 +72,7 @@ final class AddressPickerViewController: UIViewController, CheckoutProviderType
     private func showAddressListViewController(addresses: [Address]) {
         let selectedAddress: Address? = nil
 
-        let addressListViewController = AddressListViewController(addresses: addresses, selectedAddress: selectedAddress,
+        let addressListViewController = AddressListViewController(checkout: checkout, addresses: addresses, selectedAddress: selectedAddress,
```
Comment (bonkey) [inline AtlasUI/AtlasUI/AddressPickerViewController.swift]: 5 arguments in call 😱

Comment (ahmed-shehata) [inline AtlasUI/AtlasUI/AddressPickerViewController.swift]: Got to pass them all ;)

Comment (bonkey) [inline AtlasUI/AtlasUI/AddressPickerViewController.swift]: And how about removing this funky addressType. Convince me why we need it...

Comment (ahmed-shehata) [inline AtlasUI/AtlasUI/AddressPickerViewController.swift]: I'd rather not have multiple functions that do the same thing, but have types. I agree it's ugly, but there has to be a third better way..

**Comment by ahmed-shehata** [comment]
👍

**Comment by ahmed-shehata** [comment]
👍

File: AtlasUI/AtlasUI/AddressPickerViewController.swift (hunk 4 above)
Comment (bonkey) [inline AtlasUI/AtlasUI/AddressPickerViewController.swift]: Yeah, pair deleting, I mean programming...

**Comment by ahmed-shehata** [comment]
👍

**Comment by ahmed-shehata** [comment]
👍

File: AtlasDemo/AtlasDemoUITests/AtlasDemoUITests.swift [hunk 5]
```diff
@@ -16,7 +16,7 @@ class AtlasDemoUITests: XCTestCase {
 
     override class func tearDown() {
         super.tearDown()
-        try! AtlasMockAPI.stopServer() // swiftlint:disable:this force_try
+        // try! AtlasMockAPI.stopServer() // swiftlint:disable:this force_try
```
Comment (bonkey) [inline AtlasDemo/AtlasDemoUITests/AtlasDemoUITests.swift]: Why:
1. comment when it could be deleted
1. commented out when it _probably_ shouldn't?

Comment (ahmed-shehata) [inline AtlasDemo/AtlasDemoUITests/AtlasDemoUITests.swift]: Was testing something, forgot to uncoment, fixed.

File: AtlasUI/AtlasUI/AddressListTableViewDelegate.swift [hunk 6]
```diff
@@ -0,0 +1,83 @@
+//
+//  Copyright © 2016 Zalando SE. All rights reserved.
+//
+
+import Foundation
+import AtlasSDK
+
+class AddressListTableViewDelegate: NSObject {
```
Comment (bonkey) [inline AtlasUI/AtlasUI/AddressListTableViewDelegate.swift]: Why `: NSObject` and why not `: UITableViewDelegate`?

File: AtlasUI/AtlasUI/AddressListTableViewDelegate.swift [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    internal var checkout: AtlasCheckout
+    private let addressType: AddressType
+    private let selectionCompletion: AddressSelectionCompletion
```
Comment (bonkey) [inline AtlasUI/AtlasUI/AddressListTableViewDelegate.swift]: You know I'll tell again: don't use AddressType and drop it. And don't make this property required. It's not.

File: AtlasUI/AtlasUI/AddressListTableViewDelegate.swift [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+extension AddressListTableViewDelegate: UITableViewDelegate {
+
+    func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
+        return true
+    }
+
+    func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle,
+        forRowAtIndexPath indexPath: NSIndexPath) {
+            if editingStyle == .Delete {
```
Comment (bonkey) [inline AtlasUI/AtlasUI/AddressListTableViewDelegate.swift]: maybe `switch editingStyle {` instead?

File: AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift [hunk 9]
```diff
@@ -35,8 +35,10 @@ extension CheckoutSummaryActionsHandler {
             return
         }
 
-        let updateCheckoutRequest = UpdateCheckoutRequest(billingAddressId: viewController.checkoutViewModel.selectedBillingAddressId,
-            shippingAddressId: viewController.checkoutViewModel.selectedShippingAddressId)
+        let updateCheckoutRequest = UpdateCheckoutRequest(billingAddressId:
```
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]:
```
UpdateCheckoutRequest {

  convenience init(checkoutViewModel: CheckoutViewModel) {
    self.init(blllingAddressId: checkoutViewModel.selectedBillingAddress?.id,
      shippingAddressId: viewController.checkoutViewModel.selectedShippingAddress?.id)
  }
  init(checkoutViewModel: CheckoutViewModel) { ... }

}
```

File: AtlasUI/AtlasUI/AddressListTableViewDelegate.swift (hunk 6 above)
Comment (ahmed-shehata) [inline AtlasUI/AtlasUI/AddressListTableViewDelegate.swift]: Objective-C conformance to extensions (cc: @hani-ibrahim )

File: AtlasUI/AtlasUI/AddressListTableViewDelegate.swift (hunk 7 above)
Comment (ahmed-shehata) [inline AtlasUI/AtlasUI/AddressListTableViewDelegate.swift]: Will do in another issue.

File: AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift [hunk 10]
```diff
@@ -137,27 +143,27 @@ extension CheckoutSummaryActionsHandler {
...
+            viewController.refreshViewData()
             guard let ready = viewController.checkoutViewModel.isReadyToCreateCheckout where ready == true
             else { return }
             viewController.loaderView.show()
             guard let cartId = viewController.checkoutViewModel.cartId else { return }
             viewController.checkout.client.createCheckout(cartId,
-                billingAddressId: viewController.checkoutViewModel.selectedBillingAddressId,
-                shippingAddressId: viewController.checkoutViewModel.selectedShippingAddressId) { result in
+                billingAddressId: viewController.checkoutViewModel.selectedBillingAddress?.id,
+                shippingAddressId: viewController.checkoutViewModel.selectedShippingAddress?.id) { result in
```
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]: Make it just `viewController.createCheckout(cartId) { result in`. All the information you pass is already there!

Just don't use more than one object nesting and you'll be much more happier and wealthier in time saved.

File: AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift (hunk 9 above)
Comment (ahmed-shehata) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]: As I have mentioned before, you should not mix "View" model with the core SDK (AtlasUI/AtlasSDK)

Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]: I didn't mean it. You can create extension here

**Comment by ahmed-shehata** [comment]
👍

File: AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift (hunk 10 above)
Comment (ahmed-shehata) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]: Not sure what you mean? But conforming to the createCheckout API Endpoint (if there are no defaults set), you need to pass the ID's.

**Comment by bonkey** [comment]
👍

File: AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift (hunk 10 above)
Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]: I mean you should ask "My dear, would you be so kind to pass me a salt" not "My dear, would you be so kind to open your left palm of your left hand, look to your left, grab a salt with your left hand, not forgetting to close your thumb, index, middle, ring fingers and a pinky of your left palm on the container, then pass it to your right hand, then pass it to my left hand".

Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]: Or avoid mentioned already recently [feature envy](https://sourcemaking.com/refactoring/smells/feature-envy) anti-pattern.

Comment (bonkey) [inline AtlasUI/AtlasUI/CheckoutSummaryActionsHandler.swift]: And [inappropriate intimacy](https://sourcemaking.com/refactoring/smells/inappropriate-intimacy) as well.

**Pull request merged**